source/reduce/structured_loop_to_selection_reduction_opportunity.cpp

```diff
@@ -0,0 +1,406 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "structured_loop_to_selection_reduction_opportunity.h"
+#include "source/opt/aggressive_dead_code_elim_pass.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace reduce {
+
+namespace {
+const uint32_t kMergeNodeIndex = 0;
+const uint32_t kContinueNodeIndex = 1;
+}  // namespace
+
+bool StructuredLoopToSelectionReductionOpportunity::PreconditionHolds() {
+  // Is the loop header reachable?
+  return loop_construct_header_->GetLabel()
+      ->context()
+      ->GetDominatorAnalysis(enclosing_function_)
+      ->IsReachable(loop_construct_header_);
+}
+
+void StructuredLoopToSelectionReductionOpportunity::Apply() {
+  // Force computation of dominator analysis and CFG before we start to mess
+  // with edges in the function.
+  context_->GetDominatorAnalysis(enclosing_function_);
+  context_->cfg();
+
+  // (1) Redirect edges that point to the loop's continue target to their
+  // closest merge block.
```

Comment:
Great example!  No, I didn't have a test for this, and the pass gives an invalid module on the example you gave.

Rather than working out how best to handle this case, I propose adding a check for this scenario to the conditions under which we cautiously do not apply the transformation when opportunities are identified in GetAvailableOpportunities.

I will add this test as an example where there should be 0 opportunities to turn loops into selections (due to the loop being too hard).

If we find this is a blocker to reducing examples in practice we can revisit.